Provide hooks to implement and customize SQLite locking.

kentonv opened the pull request:
This allows SQLite's built-in locking to be used to protect multiple `SqliteDatabase` instances opening the same in-memory file and operating on it from different threads.

This introduces a new `SqliteDatabase::LockManager` interface which can be overridden. Arguably, an alternative could have been to extend the KJ filesystem API to support OS advisory locks. I settled on `LockManager` for a few reasons:

* In WAL mode, SQLite also does synchronization through shared memory. The VFS layer's interface to this is weird: the shared memory is treated as a facet of the database file. But in practice, the default VFS implementation opens a separate file for this. Obviously kj::File isn't going to grow an interface to hold extra "shared memory", but having to go back up to the enclosing directory in order to open a sibling file proves awkward too. With LockManager, though, it makes sense for this shared memory to be treated as part of the locking interface.

* A system which introspects on SQLite in order to, for example, implement transparent network replication, will likely need to customize this functionality. It would be awkward to try to customize by intercepting filesystem-level calls; implementing an interface specific to SQLite's higher-level operations is much more natural.

* Filesystem advisory locks have wildly different semantics and quirks across operating systems. It's not clear that KJ could offer a reasonable abstraction that covers both Unix and Windows. Even focusing on Unix, POSIX-standard locks are bizarrely connected to the calling process and not to the open file description, which has awkward effects both when one process unknowningly opens the same file multiple times, as well as when passing FDs between process. Linux has a somewhat better version of locks that are tied to open file descriptions, which would be highly preferable for our use case, but it's probably untenable for KJ to use FD-attached locks on Linux while using process-attached locks on MacOS, as there have totally different behavior. Trying to extend KJ filesystem APIs with locking would therefore present a lot of pain, but providing a separate abstraction specific to SQLite nicely sidesteps all this.

@trevorcf for review. Also cc @geelen.

trevorcf commented on src/workerd/util/sqlite.h:
```diff
@@ -256,6 +280,112 @@ class SqliteDatabase::Vfs {
+    // to `EXCLUSIVE` only when shutting down the database, in order to safely delete the WAL and
+    // WAL-index (-shm) files.
+
+    UNLOCKED,
```
not: Comment that this mirrors SQLITE_LOCK_* to avoid including .h file

trevorcf commented on src/workerd/util/sqlite.c++:
```diff
@@ -889,6 +1168,16 @@ SqliteDatabase::Vfs::Vfs(const kj::Directory& directory)
+SqliteDatabase::Vfs::Vfs(const kj::Directory& directory, const LockManager& lockManager)
+    : directory(directory),
+      lockManager(lockManager),
+      name(kj::str("kj-", &directory)),
```
nit: move "kj-" into a constant since it's used more than once

trevorcf reviewed (commented):
I still am not able to properly approve this as it says I don't have access (I believe I need to be added to terraform at this point), but I approve it

kentonv commented:
(Just a rebase on main so far.)

byule reviewed (approved):
On behalf of Trevor

kentonv commented:
Addressed review comments: https://github.com/cloudflare/workerd/compare/35617f6c90dc3ef687ec2934c924ffe6a5d6a801..7f1ce55c7b0ca40192c9e5e4f4804ef4860bf854

kentonv commented:
omgwtf, my cleanup is actually broken but only in a way that isn't covered by the tests yet. Added fix to #486.
